{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Save Sequences",
      "provenance": [],
      "collapsed_sections": [],
      "mount_file_id": "18QP8SQAI4EeVNjSSlI5KmOL73HcG1G-Y",
      "authorship_tag": "ABX9TyP29Mc/APzlb9NwVQ+7LI5B",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Yasith-Banuka/PhyloMap/blob/master/Data/Save_Sequences.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython"
      ],
      "metadata": {
        "id": "SG1-tnPLtZNA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio import Entrez\n",
        "import Bio\n",
        "import pandas as pd"
      ],
      "metadata": {
        "id": "EGEy0l2BsAfH"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "covid_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/covid_all.csv\"\n",
        ")\n",
        "\n",
        "dengue_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/dengue_all.csv\"\n",
        ")\n",
        "\n",
        "ebola_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/ebola_all.csv\"\n",
        ")\n",
        "\n",
        "enterovirus_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/enterovirus_all.csv\"\n",
        ")\n",
        "\n",
        "measles_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/measles_all.csv\"\n",
        ")\n",
        "\n",
        "mumps_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/mumps_all.csv\"\n",
        ")\n",
        "\n",
        "tb_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/tb_all.csv\"\n",
        ")\n",
        "\n",
        "zika_df = pd.read_csv(\n",
        "    \"/content/drive/MyDrive/BioS_Project/Data/zika_all.csv\"\n",
        ")\n",
        "\n",
        "dfList = [covid_df, dengue_df, ebola_df, enterovirus_df, measles_df, mumps_df, tb_df, zika_df]\n",
        "\n",
        "viruses = ['covid', 'dengue', 'ebola', 'enterovirus', 'measles', 'mumps', 'tuberculosis', 'zika']\n",
        "\n",
        "for id, df in enumerate(dfList):\n",
        "  df.name = viruses[id]"
      ],
      "metadata": {
        "id": "wxMz0NwyM7IC"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "covid_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 662
        },
        "id": "jP79_EVqNaVb",
        "outputId": "dcdb1d74-d400-4fef-eef0-36fb0c2e2675"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-4b77f665-7c67-4a67-87e0-809d4c187c33\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>strain</th>\n",
              "      <th>date</th>\n",
              "      <th>S1_mutations</th>\n",
              "      <th>clade_membership</th>\n",
              "      <th>country</th>\n",
              "      <th>current_frequency</th>\n",
              "      <th>division</th>\n",
              "      <th>emerging_lineage</th>\n",
              "      <th>epiweek</th>\n",
              "      <th>accession</th>\n",
              "      <th>host</th>\n",
              "      <th>mutational_fitness</th>\n",
              "      <th>pango_lineage</th>\n",
              "      <th>recency</th>\n",
              "      <th>region</th>\n",
              "      <th>url</th>\n",
              "      <th>author</th>\n",
              "      <th>originating_lab</th>\n",
              "      <th>submitting_lab</th>\n",
              "      <th>logistic_growth</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Wuhan-Hu-1/2019</td>\n",
              "      <td>12/26/2019</td>\n",
              "      <td>NaN</td>\n",
              "      <td>19A</td>\n",
              "      <td>China</td>\n",
              "      <td>NaN</td>\n",
              "      <td>China</td>\n",
              "      <td>19A</td>\n",
              "      <td>201952</td>\n",
              "      <td>MN908947</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>NaN</td>\n",
              "      <td>B</td>\n",
              "      <td>Older</td>\n",
              "      <td>Asia</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Wu et al</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Wuhan/WH01/2019</td>\n",
              "      <td>12/26/2019</td>\n",
              "      <td>NaN</td>\n",
              "      <td>19A</td>\n",
              "      <td>China</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Hubei</td>\n",
              "      <td>19A</td>\n",
              "      <td>201952</td>\n",
              "      <td>LR757998</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>-0.01</td>\n",
              "      <td>B</td>\n",
              "      <td>Older</td>\n",
              "      <td>Asia</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Hunter et al</td>\n",
              "      <td>BGI, Institute of Microbiology, Chinese Academ...</td>\n",
              "      <td>Beijing Genome Institute</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>CHN/P6-SARS-CoV-2/2020</td>\n",
              "      <td>1/28/2020</td>\n",
              "      <td>NaN</td>\n",
              "      <td>19A</td>\n",
              "      <td>China</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Hunan</td>\n",
              "      <td>19A</td>\n",
              "      <td>202005</td>\n",
              "      <td>MW011764</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>-0.01</td>\n",
              "      <td>B</td>\n",
              "      <td>Older</td>\n",
              "      <td>Asia</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Qiu et al</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>AUS/SA03/2020</td>\n",
              "      <td>1/30/2020</td>\n",
              "      <td>NaN</td>\n",
              "      <td>19A</td>\n",
              "      <td>Australia</td>\n",
              "      <td>NaN</td>\n",
              "      <td>South Australia</td>\n",
              "      <td>19A</td>\n",
              "      <td>202005</td>\n",
              "      <td>MT745748</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>0.00</td>\n",
              "      <td>B.1</td>\n",
              "      <td>Older</td>\n",
              "      <td>Oceania</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Caly et al A</td>\n",
              "      <td>Victorian Infectious Diseases Reference Labora...</td>\n",
              "      <td>The Peter Doherty Institute for Infection and ...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>AUS/SA04/2020</td>\n",
              "      <td>1/30/2020</td>\n",
              "      <td>NaN</td>\n",
              "      <td>19A</td>\n",
              "      <td>Australia</td>\n",
              "      <td>NaN</td>\n",
              "      <td>South Australia</td>\n",
              "      <td>19A</td>\n",
              "      <td>202005</td>\n",
              "      <td>MT745749</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>0.00</td>\n",
              "      <td>B.1</td>\n",
              "      <td>Older</td>\n",
              "      <td>Oceania</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Caly et al B</td>\n",
              "      <td>Victorian Infectious Diseases Reference Labora...</td>\n",
              "      <td>The Peter Doherty Institute for Infection and ...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4b77f665-7c67-4a67-87e0-809d4c187c33')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-4b77f665-7c67-4a67-87e0-809d4c187c33 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-4b77f665-7c67-4a67-87e0-809d4c187c33');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                   strain  ... logistic_growth\n",
              "0         Wuhan-Hu-1/2019  ...             NaN\n",
              "1         Wuhan/WH01/2019  ...             NaN\n",
              "2  CHN/P6-SARS-CoV-2/2020  ...             NaN\n",
              "3           AUS/SA03/2020  ...             NaN\n",
              "4           AUS/SA04/2020  ...             NaN\n",
              "\n",
              "[5 rows x 20 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "Entrez.email = \"yoshani.ranaweera.17@cse.mrt.ac.lk\"  \n",
        "\n",
        "for df in dfList:\n",
        "\n",
        "  !mkdir /content/drive/MyDrive/BioS_Project/Fasta_Files/{df.name}\n",
        "  print(\"------------------------> \",df.name)\n",
        "  accessionList = df['accession'].tolist()\n",
        "  print(accessionList)\n",
        "\n",
        "  for accession in accessionList:\n",
        "\n",
        "    if accession == None:\n",
        "      continue\n",
        "\n",
        "    filename = \"/content/drive/MyDrive/BioS_Project/Fasta_Files/\"+df.name+\"/\"+accession+\".fasta\"\n",
        "    if not os.path.isfile(filename):\n",
        "        try:\n",
        "          net_handle = Entrez.efetch(\n",
        "              db=\"nucleotide\", id=accession, rettype=\"fasta\", retmode=\"text\"\n",
        "          )\n",
        "          out_handle = open(filename, \"w\")\n",
        "          out_handle.write(net_handle.read())\n",
        "          out_handle.close()\n",
        "          net_handle.close()\n",
        "        except:\n",
        "          print(accession, \"not found in NCBI\") "
      ],
      "metadata": {
        "id": "OTc3tU9gtNsA"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}